<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Motivation for Reproducible Research</title>
    <meta charset="utf-8" />
    <meta name="author" content="Hitchhiker’s Guide to Reproducible Research    Julia Wrobel and David Benkeser " />
    <script src="motivation_files/header-attrs-2.26/header-attrs.js"></script>
    <link rel="stylesheet" href="xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: left, middle, inverse, title-slide

.title[
# Motivation for Reproducible Research
]
.author[
### Hitchhiker’s Guide to Reproducible Research <br> <span style="font-size: 75%;"> Julia Wrobel and David Benkeser </span>
]
.date[
### <br> <svg viewBox="0 0 512 512" style="height:1em;position:relative;display:inline-block;top:.1em;fill:#ffffff;" xmlns="http://www.w3.org/2000/svg"> <path d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"></path></svg> <a href="https://bit.ly/sismid_hitchhikers">.white[Course Website]</a>
]

---




&lt;style type="text/css"&gt;
.remark-slide-content {
    font-size: 22px
}
&lt;/style&gt;

## An email you may have received

.center[
	.monobox[
		.left[
			David, 
			&lt;br&gt;
			&lt;br&gt; After review, we found that some data may have been 
			mislabeled. Could you re-run the analyses and update numbers in 
			Tables 1-3 and re-create Figure 2?
			&lt;br&gt;
			&lt;br&gt; We would like to submit this paper by the end of the week.
			&lt;br&gt;
			&lt;br&gt; Best wishes, 
			&lt;br&gt; Your favorite collaborator
		]
	]
]

???

Very commonly in collaborations, the data change. It could be 
due to errors, more data being added, etc... The goal of this class 
is to make it so you do not fill with dread when you receive such an 
e-mail.

We are working toward a world where you download the new data,
type one command and voila! No by-hand entering numbers into tables. 
No copy-pasting from R/SAS/STATA to Word. Just one click.

---

## An email you may have received

.center[
	.monobox[
		.left[
			David, 
			&lt;br&gt;
			&lt;br&gt; The data will be finalized tomorrow and we need to have 
			new Tables to the data safety monitoring board in two days. Hope
			you did not have plans this weekend!
			&lt;br&gt;
			&lt;br&gt; Respectfully, 
			&lt;br&gt; Your boss
		]
	]
]


???


To prepare for situations like this, project organization is key. We will discuss strategies for preparing fast-turnaround projects. But just because we are turning projects around quickly, does not mean we are not working carefully. By having a set, reproducible workflow in place before data acquisition, we can have the best of both worlds. Careful analysis planning and execution AND fast delivery.

---

## An email you may have received

.center[
	.monobox[
		.left[
			David, 
			&lt;br&gt;
			&lt;br&gt; IT just updated R to the latest version on my computer and my analysis code no longer runs. I get strange error messages about compilers and package versions. My defense is next week. I am starting to panic.
			&lt;br&gt;
			&lt;br&gt; Most sincerely,
			&lt;br&gt; A frantic PhD student
		]
	]
]

???

Maintaining stable code can be challenging! We'll discuss the tools that make this easier to achieve. 

---

## A personal story

In Fall 2020, I became involved in the five US government-funded studies of COVID-19 vaccines.

__Scientific goal__: understand whether/how immune responses do/not correlate with risk of COVID-19.
* Referred to as vaccine *correlates of protection*
* Aids regulatory decisions, vaccine development, motivates updated vaccine formulations, ...

Needed a harmonized analysis across all five trials. And fast!

.red[No direct access to Moderna's data!]

---

## Reproducibility


![](motivation_files/figure-html/repro-fig-1.png)&lt;!-- --&gt;
.center[
	.small[
		Working terminology (figure produced using [scifigure](https://cran.r-project.org/web/packages/scifigure/vignettes/Visualizing_Scientific_Replication.html) R package)
	]
]

???

We will focus on building analyses that satisfy the reproducibility 
definition outlined above. Could a different analyst sit down with your code, 
execute it and get exactly the same results?

Reproducibility is a *minimal* standard. Just because something is 
*reproducible* does not necessarily imply that it is *correct*. The 
code may have bugs. The methods may be poorly behaved. But reproducibility 
is likely correlated with correctness -- if you are careful enough to 
make everything reproducible, you're probably more likely to be careful
in designing the analysis in the first place. 

We will use replicability to mean a new experiment using the same methods
coming to the same conclusions. 

Generalizability refers to taking conclusions from your study population
and generalizing them to a new population. 

---

## Other definitions of reproducibility

Our working definition of reproducibility: 

.center[.red[Could a different analyst sit down with your code, 
execute it (with minimal effort?) and get exactly the same results?]]

Many perspectives on reproducibility/replicability/etc... Some readings:
- [Peng *Reproducible Research in Computational Science*](https://www.science.org/doi/10.1126/science.1213847)
- [Patil, et al. *A statistical definition of reproducibility*](https://www.biorxiv.org/content/biorxiv/early/2016/07/29/066803.full.pdf)
- [Goodman, et al. *What does research reproducibility mean?*](https://www.science.org/doi/10.1126/scitranslmed.aaf5027)

---
class: title-slide, center, inverse, middle
background-color: #84754e

*You mostly collaborate with yourself, and me-from-two-months-ago never responds to email.*
&lt;div&gt; - Karen Cranston

---

## Why should I care?

* Careful coding means more likely to produce __correct results__.
* In the long run, you will __save (a lot of) time__.
* Higher __impact__ of your science.
* __Avoid embarrassment__ on a public stage. 

???

Doing reproducible science is time consuming, but is ultimately time-saving.
As you practice these tools more and more, they will start to become
habits. I do analyses now in a day or so (and at much higher quality) now
that would take me a (frustrating) week seven or eight years ago. 

One of my great fears as an academic is being accused of intentionally 
doing bad science. Leaving a paper trail that any one else can follow is
a bit intimidating -- we all make mistakes. Reproducible work indicates 
that any mistakes made were likely honest.


---

## Odyssey of Baggerly and Coombes

A prominent [reproducibility horror story](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3474449/) occurred in the late 20-00's at Duke University Institute for Genome Sciences and Policy.

Team was using __genomic signatures__ to predict the __most effective chemo regimen__ for cancer patients.

Keith Baggerly and Kevin Coombes (MD Anderson) .red[raised questions about the validity] of the results.

After Duke was alerted that the PI of trial lied on his CV about being a Rhodes scholar(!), an investigation was launched that uncovered:
* data management errors 
* incorrectly designed computational models
* failure to validate gene expression tests 
* what appeared to be intentional data corruption

---

class: title-slide, center, inverse, middle
background-color: #828280

*This is not rocket science. There is computer code that evaluates the algorithm. There is data. And when you plug the data into that code, you should be able to get the answers back that you have reported. And to the extent that you can’t do that, there is a problem in one or both of those items.*
&lt;div&gt; - Lisa McShane

---

## Levels of reproducibility

* Are the tables and figures reproducible using the code?
* Are the tables and figures reproducible using the code __with minimal user effort__?
* Does the code __actually do__ what's written in Methods?
* Does the code make clear __why__ certain decisions were made?
	* Why did you exclude this variable? 
	* Why do you choose these tuning parameters?
* Can the code be used on other data?
* Can the code be extended to do more?

???

Reproducibility is not black and white. The ideal is not easy to 
achieve, nor is it necessarily warranted on every project. This is a lifelong
process of understanding what level of reproducibility is required for 
each project. 

---

## Basic principles

* Get data in its "raw-est" form possible
* __Document__ where/when you got the data
* Be as __self-sufficient__ as possible
* __Everything via code__
	* with sufficient documentation!
* __Everything automated__
	* with sufficient documentation! 

???

The goal is end-to-end reproducibility. In many situations, 
*you* will be the person with the most data 
management experience. Reproducibility can easily be undermined
by e.g., a collaborator who is manually extracting data from an 
online data base and copy-pasting it into a .csv for you. 

Solution: Code the download process! Automate data pulls! Document
how you're doing it!

---

## Things to avoid

* Open a file to re-save as `.csv`. 
* Open a data file and manually edit. 
* Pasting results into the tables for a manuscript.
* Copy-paste-edit code into GUI
* Copy-paste-edit tables
* Copy-paste-adjust figures (with due respect to MS Paint)

???

All of these actions are problematic. Each introduces the possibility
of inconsistencies or errors and is not reproducible in the sense 
that we mean. 

---
class: title-slide, center, inverse, middle
background-color: #007dba

*If you do anything "by hand" once, you'll have to do it 100 times.*
&lt;div&gt; - Karl Broman

???

So true. 

---

## Tools for reproducible research

* Unix command line
* Scripting languages (e.g., `bash`, `zsh`)
* Markdown, RMarkdown, Latex, Knitr
* Version control (e.g., git and GitHub)
* GNU Make
* R packages
* Containers (e.g., Docker)

???

In this class, we'll be learning to operate almost exclusively on 
the command line. Soon, [this will be you](https://tenor.com/8IZG.gif). 

Git will make your life easier by handling version control. No more
directories with 200 copies of the same file with different dates!

RMarkdown + knitr are great tools for generating reproducible reports (or
course notes...). 

GNU Make is for automation and for documenting dependencies between 
files. 

`R` packages are a useful way to wrap up code, documentation, and dependencies
on other `R` packages. 

Containers make it possible to achieve full reproducibility rather 
painlessly, by wrapping all needed software (down to the OS!) up into a 
"container". 

Knowing some `bash` is necessary to operate at the command line and 
generally useful for scripting. `R` is becoming (or already is) a standard 
software for analysis. `python` is also becoming quite popular for analysis, 
but goes much beyond. Other languages are great, but not totally necessary.

---
class: title-slide, center, inverse, middle
background-color: #487f84

*The most important tool is the mindset, when starting, that the end product will be reproducible.*
&lt;div&gt; - Keith Baggerly


    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create();
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
